### PR TITLE
Extract curve25519 crate from zk-token-sdk

### PR DESCRIPTION
**Pull request by kevinheavey**
#### Problem
As the zk-token-sdk crate itself correctly notes, the code in the curve25519 folder should live in its own crate:
```
//! This module lives inside the zk-token-sdk for now, but should move to a general location since
//! it is independent of zk-tokens.
```
A more pressing motivation for me is that I have code that depends on bpf_loader, which only needs the curve25519 code from zk-token-sdk, and it takes a while to compile the whole zk-token-sdk crate.

#### Summary of Changes
Extracts `zk-token-sdk/src/curve25519` into its own crate and re-exports it for backwards compatibility. Updates code in the monorepo to use the new crate directly.

### CODE REVIEW

**Comment by kevinheavey** [comment]
Think the downstream checks are failing just because the new crate isn't on crates.io yet

**Comment by Xuanwo** [comment]
Hi, should we merge https://github.com/anza-xyz/agave/pull/946 first before doing this refactor?

**Comment by kevinheavey** [comment]
When that's merged I will rebase and make sure the relevant changes to the curve25519 dir are ported over

**Comment by samkim-crypto** [comment]
@kevinheavey, I wanted to merge https://github.com/anza-xyz/agave/pull/1495 and https://github.com/anza-xyz/agave/pull/1693 in before this PR, but it seems like those PRs are getting delayed and it would be ideal to get this PR in before agave 2.0 is cut. Do you mind rebasing the PR to test the CI? And let's put this crate under the directory `curves` as was done in https://github.com/anza-xyz/agave/pull/1495.

File: curves/curve25519/src/lib.rs [hunk 1]
```diff
@@ -1,7 +1,5 @@
+#![allow(clippy::arithmetic_side_effects, clippy::op_ref)]
 //! Syscall operations for curve25519
-//!
-//! This module lives inside the zk-token-sdk for now, but should move to a general location since
-//! it is independent of zk-tokens.
 
 pub mod curve_syscall_traits;
 pub mod edwards;
```
Comment (samkim-crypto) [inline curves/curve25519/src/lib.rs:5]: Can we get rid of `#[cfg(not(target_os = "solana"))]` in the line below? Sorry, github interface does not allow me to comment on the exact line.

**Review by samkim-crypto — COMMENTED** [review]
Okay thanks! I think the CI is not too bad. I can reserve the crate to take care of `crate-check`. Other than that, it appears that downstream tests are the main ones.

Getting rid of the `target_os` restriction below will take care of some failures, so let's make that change and see what we get with the CI.

**Comment by samkim-crypto** [comment]
Sorry for the delay on this 🙏 . All the CI passes, so I will merge it after you rebase.

**Review by samkim-crypto — APPROVED** [review]
Thank you for the changes!

**Pull request merged**